pkg/api/meta/priority.go

```diff
@@ -0,0 +1,173 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package meta
+
+import (
+	"fmt"
+
+	"k8s.io/kubernetes/pkg/api/unversioned"
+)
+
+const (
+	AnyGroup    = "*"
+	AnyVersion  = "*"
+	AnyResource = "*"
+	AnyKind     = "*"
+)
+
+// PriorityRESTMapper is a wrapper for automatically choosing a particular Resource or Kind
+// when multiple matches are possible
+type PriorityRESTMapper struct {
+	// Delegate is the RESTMapper to use to locate all the Kind and Resource matches
+	Delegate RESTMapper
+
+	// ResourcePriority is a list of priority patterns to apply to matching resources.
+	// The list of all matching resources is narrowed based on the patterns until only one remains.
+	// A pattern with no matches is skipped.  A pattern with more than one match uses its
+	// matches as the list to continue matching against.
+	ResourcePriority []unversioned.GroupVersionResource
```

Comment:
> I'm not sure that's right. They need higher priority ones first, but less specificity does make sense in some cases. For instance, if your example had A, B, and C, but in the case of resource R, I want A/R, C/R, B/R, then I still logically want A (less specific) to come first.

Reflecting on the comment above, I think we need the following 2 rules:
1. if user specifies a special case for a resource, he needs to specify the patterns for all the possible groups. In the example used in the comment above, user needs to specify 3 patterns: A/R, C/R, B/R, (instead of only C/R, B/R);
2. more specific rules need to come first in the list. So in the example above, the merged rules will be (A/R, C/R, B/R, A, B, C)

Without rule 1, you will have ambiguity when merging user config with discover doc.